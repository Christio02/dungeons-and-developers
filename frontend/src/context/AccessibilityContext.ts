import { createContext, useContext } from 'react';

export const AccessibilityContext = createContext({
  isAccessibilityMode: false,
  toggleAccessibilityMode: () => {},
});
export const useAccessibilityContext = () => {
  const context = useContext(AccessibilityContext);
  if (!context) {
    throw new Error('useCharacterContext must be used within a CharacterProvider');
  }
  return context;
};
