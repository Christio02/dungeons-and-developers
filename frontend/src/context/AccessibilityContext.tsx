import React, { createContext, useContext, useEffect, useState } from 'react';

const AccessibilityContext = createContext({
  isAccessibilityMode: false,
  toggleAccessibilityMode: () => {},
});

export const AccessibilityProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [isAccessibilityMode, setIsAccessibilityMode] = useState(false);
  const [isInitialized, setIsInitialized] = useState(false);

  useEffect(() => {
    const storedMode = localStorage.getItem('accesibility-mode') === 'true';
    setIsAccessibilityMode(storedMode);
    setIsInitialized(true);
  }, []);

  const toggleAccessibilityMode = () => {
    setIsAccessibilityMode((prev) => !prev);
  };

  useEffect(() => {
    if (isInitialized) {
      if (isAccessibilityMode) {
        localStorage.setItem('accesibility-mode', 'true');
        document.body.classList.add('accessibility-mode');
      } else {
        localStorage.setItem('accesibility-mode', 'false');
        document.body.classList.remove('accessibility-mode');
      }
    }
  }, [isAccessibilityMode, isInitialized]);

  return (
    <AccessibilityContext.Provider value={{ isAccessibilityMode, toggleAccessibilityMode }}>
      {children}
    </AccessibilityContext.Provider>
  );
};

export const useAccessibility = () => useContext(AccessibilityContext);
