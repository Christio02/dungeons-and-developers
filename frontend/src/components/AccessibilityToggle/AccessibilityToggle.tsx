import React from 'react';
import { useLocation } from 'react-router-dom';
import accessabilityIcon from '../../assets/images/accessibility-icon.png';
import { useAccessibilityContext } from '../../context/AccessibilityContext.ts';
import AccesibilityProps from '../../interfaces/AccesibilityProps.ts';
/**
 * AccessibilityToggle component allows users to change the complex UI to a more accessible mode.
 *
 * Renders a component that user can interact with as an image and a checkbox.
 */
const AccessibilityToggle = ({ checked = false, onChange }: AccesibilityProps) => {
  const { isAccessibilityMode, toggleAccessibilityMode } = useAccessibilityContext();
  const location = useLocation();
  const isLoginPage = location.pathname === '/';

  const handleCheckboxChange = () => {
    const newChecked = !isAccessibilityMode;
    toggleAccessibilityMode();
    if (!checked) {
      onChange?.(newChecked);
    }
  };
  const handleKeyDown = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter' || event.key === ' ') {
      handleCheckboxChange();
    }
  };

  return (
    <label className="flex items-center cursor-pointer z-50" onKeyDown={handleKeyDown} tabIndex={0}>
      <div className="mr-3">
        <img
          src={accessabilityIcon}
          alt="accessability icon"
          className={`w-10 ${isLoginPage ? 'filter invert' : ''}`}
        />
      </div>
      <input
        type="checkbox"
        checked={isAccessibilityMode}
        onChange={toggleAccessibilityMode}
        className="sr-only"
        onKeyDown={handleKeyDown}
      />
      <span
        className={`relative inline-block w-14 h-8 rounded-full transition-colors ${
          isAccessibilityMode ? 'bg-[#E4BF36]' : 'bg-gray-400'
        }`}
      >
        <span
          className={`absolute top-1 left-1 w-6 h-6 bg-white rounded-full shadow-md transform transition-transform ${
            isAccessibilityMode ? 'translate-x-6' : ''
          }`}
        ></span>
      </span>
    </label>
  );
};

export default AccessibilityToggle;
