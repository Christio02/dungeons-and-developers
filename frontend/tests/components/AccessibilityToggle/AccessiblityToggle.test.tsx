import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, it, expect, vi, MockedFunction } from 'vitest';
import { useLocation } from 'react-router-dom';
import { useAccessibility } from '../../../src/context/AccessibilityContext.tsx';
import AccessibilityToggle from '../../../src/components/AccessibilityToggle/AccessibilityToggle.tsx';

// Mock the AccessibilityContext hook
vi.mock('../../../src/context/AccessibilityContext.tsx', () => ({
  useAccessibility: vi.fn(),
}));

vi.mock('react-router-dom', () => ({
  useLocation: vi.fn(),
}));

describe('AccessibilityToggle', () => {
  const mockToggleAccessibilityMode = vi.fn();

  beforeEach(() => {
    vi.resetAllMocks();
    // Typecast useAccessibility to a mocked function
    (useAccessibility as MockedFunction<typeof useAccessibility>).mockReturnValue({
      isAccessibilityMode: false,
      toggleAccessibilityMode: mockToggleAccessibilityMode,
    });
    (useLocation as MockedFunction<typeof useLocation>).mockReturnValue({
      hash: '',
      key: '',
      search: '',
      state: undefined,
      pathname: '/some-page',
    });
  });

  it('renders correctly and matches snapshot', () => {
    const { container } = render(<AccessibilityToggle />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the icon with inverted filter on the login page', () => {
    (useLocation as MockedFunction<typeof useLocation>).mockReturnValue({
      hash: '',
      key: '',
      search: '',
      state: undefined,
      pathname: '/',
    });

    render(<AccessibilityToggle />);

    const icon = screen.getByAltText(/accessability icon/i);
    expect(icon).toHaveClass('filter invert');
  });

  it('displays the icon without inverted filter on other pages', () => {
    render(<AccessibilityToggle />);

    const icon = screen.getByAltText(/accessability icon/i);
    expect(icon).not.toHaveClass('filter invert');
  });

  it('displays the toggle with the correct background color when accessibility mode is off', () => {
    render(<AccessibilityToggle />);
    const toggle = screen.getByRole('checkbox', { hidden: true }); // `hidden: true` allows us to query hidden inputs
    expect(toggle.nextSibling).toHaveClass('bg-gray-400'); // The span element with the background color
  });

  it('displays the toggle with the correct background color when accessibility mode is on', () => {
    (useAccessibility as MockedFunction<typeof useAccessibility>).mockReturnValue({
      isAccessibilityMode: true,
      toggleAccessibilityMode: mockToggleAccessibilityMode,
    });

    render(<AccessibilityToggle />);
    const toggle = screen.getByRole('checkbox', { hidden: true });
    expect(toggle.nextSibling).toHaveClass('bg-[#E4BF36]');
  });

  it('handles toggle switch change', async () => {
    render(<AccessibilityToggle />);
    const toggle = screen.getByRole('checkbox', { hidden: true });

    await userEvent.click(toggle);
    expect(mockToggleAccessibilityMode).toHaveBeenCalledTimes(1);
  });

  it('displays the toggle knob at the correct position when accessibility mode is off', () => {
    render(<AccessibilityToggle />);
    const knob = screen.getByRole('checkbox', { hidden: true }).nextSibling?.firstChild;
    expect(knob).not.toHaveClass('translate-x-6');
  });

  it('displays the toggle knob at the correct position when accessibility mode is on', () => {
    (useAccessibility as MockedFunction<typeof useAccessibility>).mockReturnValue({
      isAccessibilityMode: true,
      toggleAccessibilityMode: mockToggleAccessibilityMode,
    });

    render(<AccessibilityToggle />);
    const knob = screen.getByRole('checkbox', { hidden: true }).nextSibling?.firstChild;
    expect(knob).toHaveClass('translate-x-6');
  });
});
