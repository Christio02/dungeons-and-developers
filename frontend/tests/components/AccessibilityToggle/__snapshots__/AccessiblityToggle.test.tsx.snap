// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AccessibilityToggle > renders correctly and matches snapshot 1`] = `
<label
  class="flex items-center cursor-pointer z-50"
>
  <div
    class="mr-3"
  >
    <img
      alt="accessability icon"
      class="w-10 "
      src="/project2/src/assets/images/accessibility-icon.png"
    />
  </div>
  <input
    class="hidden"
    type="checkbox"
  />
  <span
    class="relative inline-block w-14 h-8 rounded-full transition-colors bg-gray-400"
  >
    <span
      class="absolute top-1 left-1 w-6 h-6 bg-white rounded-full shadow-md transform transition-transform "
    />
  </span>
</label>
`;
